### PR TITLE
dbAuth: Adds a totally self-contained database-based authentication option

### PR DESCRIPTION
**Pull request by cannikin**
# THE PROPHECY IS FULFILLED

The day has finally come. After many tears and gnashing of teeth, an alternative to third-party authentication has finally come. I call it: dbAuth. You store your own credentials, you log in your own users (for as long or as little time as you want), and it requires no third party integration. Your app stays self-contained, fast and awesome. There are lots of configuration options, so you're not locked into a single all-or-nothing proposition.

## Why

I've got a database. I can store my own user data (securely). I can host my own login/signup pages. I shouldn't *have* to give someone my user data just to let someone log in. I shouldn't have learn yet another third party's API/docs. One fewer third party integration means one fewer potential failure points. Fewer, faster network requests. A wise man once said that every dependency you add doesn't double the complexity of your app, it expontentializes it.

And finally, it should be part of the Redwood Way—just like deployment, we'll provide you with some of the best options (including self-hosting) and you pick what works for you.

## Background

I don't know if I've mentioned this in the past, but before Redwood I was a Ruby on Rails developer. Back in my day all you could do was your own database-based authentication, and it worked great! Simple and fast.

The experience that finally made me bite the bullet was a confluence of two events: trying to move from Netlify hosting to Render, and Netlify's token expiring and logging me out of my site every couple of hours. There was a PR meant to fix the token expiration issue, but it kept happening. As for moving providers, I was using Netlify Identity which, it turns out, is only supported by hosting your site on Netlify. So I looked into Auth0. Their documentation and installation attempt made me want to stop programming. 

I finally realized that I had be lying to myself, trying to convince myself that third party authentication was the only viable option in this day and age. Enough. There is a better way, and I already had the tools I needed to make that way a reality—my hands and my mind.

## One Function to Rule Them All

The "client" in dbAuth is just a function on the API side. And this function doesn't really have any logic, it's just creating an instance of the new internal Redwood `DbAuthHandler` class, but passing the constructor a bunch of options to configure it for your app.

This api function will look something like:

```javascript
import { db } from 'src/lib/db'
import { DbAuthHandler } from '@redwoodjs/api'

export const handler = async (event, context) => {
  const authHandler = new DbAuthHandler(event, context, {
    // Provide prisma db client
    db: db,

    // The name of the function you'd call on `db` to access your user table
    // ie. if your table is named `User` then this value would be `user`, as in `db.user`
    authModelAccessor: 'user',

    // A map of what dbAuth calls a field to what your database calls it.
    // `id` is whatever column you use to uniquely identify a user (something like `id` or even `email`)
    authFields: {
      id: 'id',
      username: 'email',
      hashedPassword: 'hashedPassword',
      salt: 'salt',
    },

    // Whatever you want to happen to your data on new user signup.
    // Redwood will check for duplicate usernames before calling this handler.
    signupHandler: ({ username, hashedPassword, salt, userAttributes }) => {
      // If you allow additional data to be included besides username and password,
      // be careful to check that only allowed data in `userAttributes` is given to Prisma:
      // a hacker could try to stuff additional data into the signup POST and sneak it into your DB

      if (!username.match(/.+?@.+?\..+?/)) {
        throw new Error('Not a valid email address!')
      }

      return db.user.create({
        data: {
          email: username,
          hashedPassword: hashedPassword,
          salt: salt,
          name: userAttributes.name,
        },
      })
    },

    // How long a user will remain logged in, in seconds
    loginExpires: 60 * 60 * 24 * 365 * 10,
  })

  return authHandler.invoke()
}

```

## Database Requirements

You need three fields on your users table, named whatever you want. They store your username (probably an email), a hashed password, and a salt. The salt is generated automatically on signup and saved with the new user. The hashed password is created by PBKDF2 hashing the user's plain text password along with their salt.

## Login Process

On login, the user's username and password are POSTed to the auth function. If a user is found with the given username, their salt is combined with their submitted password and PBKDF2 hashed. If the resulting string matches the hashedPassword in their user record, then they're authenticated. A cookie is set in the browser containing only the user's ID, AES encrypted with a `SESSION_SECRET` that is set in `.env`. The cookie is marked `HttpOnly` so it's not viewable by Javascript (which prevents XSS attacks), `Secure` (prevents man-in-the-middle attacks), `SameSite=strict` (which prevents a subset of CSRF attacks).

To verify the user is logged in, on a `getToken()` call (from AuthProvider) the function will check for the presence of the cookie, decrypt it, and look up the user with a matching ID. That ID is returned from the call (we don't do anything with it currently, but we have to return *something* because of the way the auth package takes whatever is there and puts it in an `Authorization` header when making GraphQL requests). When a GraphQL call is made, the cookies go up with it and there is a new dbAuth decoder that decrypts the cookie and returns the data inside, making it available as the first argument to `getCurrentUser()` in `api/src/lib/auth.js`. You then look up the user yourself by whatever mechanism you want.

## CSRF Tokens

This code also provides for CSRF token verification, but we'll have to modify our GraphQL engine to use it. The idea is that on each request a new CSRF token is generated and returned in two places—plain text in an `X-CSRF-Token` header, and also encrypted into the session cookie. When any POST is made to the server the previously received token should be sent in a header. That token is then compared to the one in the session and the request is only allowed to proceed if the two match. Before the response is sent, a new CSRF token is created and returned and the process repeats. [Read more about CSRF attacks](https://medium.com/swlh/csrf-cross-site-request-forgery-attack-and-ways-to-combat-it-in-rails-7d0ae7de57d2).

Ideally we would provide something like this for Redwood in general, so this functionality may be able to be extracted and generalized so it can be used with any (or no) auth provider.

## `yarn rw setup auth dbAuth`

This command creates the auth function and provides some help text for what else you need to provide (like adding `hashedPassword` and `salt` to your user table). It also creates a SESSION_SECRET variable and either creates or appends it to `.env` with a message letting you know to never check it into version control (in addition, `.env` is already included in new projects' `.gitignore` and has been since v0.1 I think).

You'll need to configure the auth function with those details specific to your app. The `signupHandler` in the auth function lets you define any logic you need to perform on user signup (like verifying that additional user data fields are correct and a sketchy client isn't trying to inject some `connect` key to link you a record they're not supposed to be connected to).

Docs produced when the command completes:

```
Done! But you have a little more work to do:

You will need to add a couple of fields to your User table in order
to store a hashed password and salt:

  model User {
    id             Int @id @default(autoincrement())
    email          String  @unique
    hashedPassword String   // <─┐
    salt           String   // <─┴─ add these lines
  }

If you already have existing user records you will need to provide
a default value or Prisma complains, so change those to:

  hashedPassword String @default("")
  salt           String @default("")

You'll need to let Redwood know what field you're using for your
users' `id` and `username` fields. In this case we're using `id` and
`email`, so update those in the `authFields` config in
`/api/src/functions/auth.js` (this is also the place to tell Redwood if
you used a different name for the `hashedPassword` or `salt` fields):

  authFields: {
    id: 'id',
    username: 'email',
    hashedPassword: 'hashedPassword',
    salt: 'salt',
  },

Finally, we created a SESSION_SECRET environment variable for you in
/Users/rob/Sites/coincontrol.dev/.env. This value should NOT be checked
into version control and should be unique for each environment you
deploy to. If you ever need to log everyone out of your app at once
change this secret to a new value. To create a new secret, run:

  yarn rw g secret

Need simple Login and Signup pages? We've got a generator for those
as well:

  yarn rw g scaffold dbAuth
```

## `yarn rw g scaffold dbAuth`

This command creates basic login/signup pages which users can then modify. They match the styling of our scaffold generators, and even use the same stylesheet and classes. Devs could also create their own pages from scratch—just call `logIn` and `signUp` (destructured from `useAuth()`) passing the needed fields to log someone in or sign them up.

There's currently no provision for a **Forgot Password** flow. Traditionally you'd get an email with a link to reset your password, but folks may not have a mail server set up (and if they did, the options vary widely). I'll need to think more about this, but my first instinct is to add a `forgotPasswordHandler()` similar to the `signupHandler()` above and just let you configure your own logic for how to get the link to the user. 

Docs output when the command completes:

```   
Pages created! But you're not done yet:

You'll need to tell your pages where to redirect after a user has logged in
or signed up. Look in LoginPage and SignupPage for these lines:

  if (isAuthenticated) {
    navigate(routes.home())
  }

and change the route to where you want them to go.

Oh, and if you haven't already, add the necessary dbAuth functions and
app setup by running:

  yarn rw setup auth dbAuth

Happy authenticating!
```

## `yarn rw secret`

This new command creates a 64-character string of random numbers/letters using a cryptographically-secure source of entropy. The `yarn rw setup auth dbAuth` command will use the same function internally to automatically create a `SECURE_SESSION` environment variable that is used to encrypt/decrypt the session cookie. If you ever want to change this value, this CLI command gives you a new one. Note that this command doesn't insert it into a file for you, it just outputs to the screen (with a message telling you how to use it with dbAuth).

## Additional Benefits

As far as I know, this is the only auth solution we have that gives you the ability to log out all users if you wish—changing the `SESSION_SECRET` and re-deploying will attempt to decrypt any users' cookies using a key that doesn't match the one it was encrypted with. This will return a "logout" response from the server which deletes the cookie and the user will be redirected to the login page.

## In Conclusion

I think this is a great option for many people and as a side benefit will greatly simplify the process for those just getting started (specifically the [Authentication](https://learn.redwoodjs.com/docs/tutorial/authentication) section of the tutorial, which is definitely where things start getting gnarly).

dbAuth: let it cleanse your soul.

image

### CODE REVIEW

**Comment by cannikin** [comment]
Shoutout to @clairefro for a Twitch stream one evening where she worked on adding cookie-based auth to a Redwood app, which finally triggered the right neurons in my brain to attempt this!

image

**Comment by peterp** [comment]
This is awesome! Since I don't host on serverless I've always wanted a simple/ controllable/ understandable mechanism to authenticate users.

Thanks for getting this started @cannikin! I know people bemoan "rolling your own auth," but I think there's a difference if a framework provides it... so that _you the user of the framework_ don't have roll your own auth.

**Comment by mojombo** [comment]
Love the idea of having a self-operated auth solution available. Let's try to get it in the 1.0 release!

**Comment by cannikin** [comment]
`yarn rw setup auth dbAuth` is looking good, here's the current help text I provide afterwords which walks you though adding a couple of fields to your user model:

```
% yarn rw setup auth dbAuth
yarn run v1.22.10
$ /Users/rob/Sites/coincontrol.dev/node_modules/.bin/rw setup auth dbAuth
  ✔ Generating auth lib...
    ✔ Successfully wrote file `./api/src/lib/auth.js`
    ✔ Successfully wrote file `./api/src/functions/auth.js`
  ✔ Adding auth config to web...
  ✔ Adding auth config to GraphQL API...
  ✔ Adding required web packages...
  ✔ Installing packages...
  ✔ One more thing...

    You will need to add a couple of fields to your User table in order
    to store their hashed password and salt values. You can name them
    something custom if you wish, just be sure to update the `authFields`
    option in `api/src/functions/auth.js` to let Redwood know what you
    called them:
   
      model User {
        id              Int @id  @default(autoincrement())
        email           String   @unique
        hashedPassword  String   // <--- add this line
        salt            String   // <--- and this one
      }
   
    If you already have existing user records you will need to provide
    a default value or Prisma complains, so change those to:
   
      hashedPassword  String   @default("")
      salt            String   @default("")
   
    Need simple Login and Signup pages? We've got a generator for those
    as well:
   
      yarn rw g scaffold dbAuth

✨  Done in 38.84s.
```

**Comment by peterp** [comment]
@cannikin I know these globals make this easier, but I'm worried about the security implications. I don't hate classes for things that make sense, and maybe this is one of them if you don't want to pass around those arguments to various functions?

**Comment by cannikin** [comment]
> @cannikin I know these globals make this easier, but I'm worried about the security implications. I don't hate classes for things that make sense, and maybe this is one of them if you don't want to pass around those arguments to various functions?

YES I'll work on that today!

**Comment by dthyresson** [comment]
On `excludeUserFields` I see that they'll be excluded from `currentUser`, but if these fields on on the `User` model/table, how do you suggest protecting these from being leaked by a GraphQL query that requests then from `users` or `user`?

```
    // Any db fields in your user table that you want to exclude from the
    // `currentUser` data response. `hashedPassword` and `salt` are always excluded
    excludeUserFields: [],
```

There might be a way with the new envelop GraphQL features to redact these such as the `usePayloadFormatter` that "allow[s] you to format/modify execution result payload before returning it to your consume": https://github.com/dotansimha/envelop/blob/8bbe150269b30fe7e48fa27e391a3484c3f8952d/packages/core/README.md#usepayloadformatter

**Comment by dthyresson** [comment]
While Prisma doesn't support querying another db schema, did you consider keeping the sensitive data like the hashed password in `auth.users` that joins your User model in Prisma (ie public.users)?

This would reduce the risk of leaking that sensitive info via GraphQL. Your query to check the password would have to use Prisma's `queryRaw` (https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access/#queryraw) to access `auth.users`. This auth users would have a `userId` that links to public.users.

You could shield sensitive profile data there as well, such as `email` such that someone count not query all users and get all email addresses.

**Comment by cannikin** [comment]
> Storing the JWT in localStorage (which pretty much all the providers do), isn't recommended.

Hmm, I'm not storing the JWT anywhere right now—it's returned fresh from the auth function on every call to `getToken`. The only thing being stored is the encrypted session cookie.

Storing the JWT in LocalStorage has made for many confusing dev sessions. I'm working on an app, then leave and work on another app. I'm now also logged into that app because the JWT from my first session is in LocalStorage and the second app picks it up and is happy to consider me authenticated (on the web side, at least)! That always felt extremely dangerous (although I guess it wouldn't be an issue in production where you have different domains and thus different LocalStorage instances).

It feels much more secure to double-check with the server on every request (web and api) whether I'm actually logged in or not. And maybe that does actually happen with our other providers, but that scenario above makes me question whether it's actually the case...

**Comment by dthyresson** [comment]
> Storing the JWT in LocalStorage has made for many confusing dev sessions. I'm working on an app, then leave and work on another app. I'm now also logged into that app because the JWT from my first 

Agreed.

> It feels much more secure to double-check with the server on every request (web and api) whether I'm actually logged in or not. 

So, every `useAuth` and every GraphQL request would also make a function call to the auth handler to return accessToken which then `auth.js` will decode and verify access? So, doe that mean that effectively each GraphQL request will x2 the function calls?

**Comment by cannikin** [comment]
> While Prisma doesn't support querying another db schema, did you consider keeping the sensitive data like the hashed password in auth.users that joins your User model in Prisma (ie public.users)?

I'm not familiar with `auth.users`... is that a Postgres thing? I don't think you can have periods in MySQL table names (or I've never seen it, at least). Not sure about SQLite.

I thought about letting you specify a separate table for storing credentials (maybe your user data is in a `User` table and your auth data in `Credentials`) but wanted to keep it simple for this first pass. Right now I'd rather it be simple and accessible for newbies than to worry about enterprise-grade customers. Unless we think that on Day 1 a large number of users are going to demand this functionality or refuse to use Redwood altogether. I'd rather start simple and add this feature later than to include even more options in the auth.js function (I'm not super happy with how many are there now, but I feel like it's the minimum we need to launch).

I wasn't thinking about GraphQL leaking info and figured that was a separate concern (which data you expose via your API) but with the generators it makes it almost *too* easy to leak info since we assume all fields are sharable. I could add a check in the SDL generator to exclude any field named `hashedPassword` or `salt` automatically, and since those are the defaults most people will probably keep those exact names. Or it could get even fancier and if you have the `auth.js` function in place, somehow read the config and look up the names you've given to `hashedPassword` and `salt` and exclude those. Although those are parameters passed to a function call so I'd have to read the whole file as text and try to parse them out... argh!

**Comment by dthyresson** [comment]
You may want to consider some additional claims here:

https://github.com/redwoodjs/redwood/blob/c231f50646b76c181052e76e8662d6061e7cf22e/packages/api/src/functions/dbAuthHandler.js#L406

```js
const token = jwt.sign(JSON.stringify(user), process.env.SESSION_SECRET)
```

Perhaps to have a very short-lived `exp` (if in fact it is to be requested frequently) and maybe an `iss` issuer?

```
iss (issuer): Issuer of the JWT

sub (subject): Subject of the JWT (the user)

aud (audience): Recipient for which the JWT is intended

exp (expiration time): Time after which the JWT expires

nbf (not before time): Time before which the JWT must not be accepted for processing

iat (issued at time): Time at which the JWT was issued; can be used to determine age of the JWT

jti (JWT ID): Unique identifier; can be used to prevent the JWT from being replayed (allows a token to be used only once)
```

The `jti` might be useful if it is in fact per request.

> I'm not familiar with `auth.users`... is that a Postgres thing? I don't think you can have periods in MySQL table names (or I've never seen it, at least). Not sure about SQLite.

Ah, yes, schemas would be a Postgres feature: https://www.postgresql.org/docs/13/ddl-schemas.html#DDL-SCHEMAS-PUBLIC

Schemas have owners and separate permissions and organize sets of tables, functions, triggers, permissions, etc. each schema is kind alike a mini db.

Since Prisma connects to Postgres with (likely) an admin or owner account (so it can delete, etc) in can access public and non public schemas.

But, Prisma's client with `findMany` etc can only access tables in the `public` schema. Thus it's not easy to access tables in another schema like `auth` because the "Prisma schema definition" is only for "public".

> By default, users cannot access any objects in schemas they do not own. To allow that, the owner of the schema must grant the USAGE privilege on the schema. To allow users to make use of the objects in the schema, additional privileges might need to be granted, as appropriate for the object.

So, if you want in Prisma to access content from a non-public schema, you have to use `queryRaw`.

**Comment by dthyresson** [comment]
> I'm not familiar with `auth.users`... is that a Postgres thing? I don't think you can have periods in MySQL table names (or I've never seen it, at least). Not sure about SQLite.

Ah, yes, schemas would be a Postgres feature: https://www.postgresql.org/docs/13/ddl-schemas.html#DDL-SCHEMAS-PUBLIC

Schemas have owners and separate permissions and organize sets of tables, functions, triggers, permissions, etc. each schema is kind alike a mini db.

Since Prisma connects to Postgres with (likely) an admin or owner account (so it can delete, etc) in can access public and non public schemas.

But, Prisma's client with `findMany` etc can only access tables in the `public` schema. Thus it's not easy to access tables in another schema like `auth` because the "Prisma schema definition" is only for "public".

> By default, users cannot access any objects in schemas they do not own. To allow that, the owner of the schema must grant the USAGE privilege on the schema. To allow users to make use of the objects in the schema, additional privileges might need to be granted, as appropriate for the object.

So, if you want in Prisma to access content from a non-public schema, you have to use `queryRaw`.

**Comment by cannikin** [comment]
> On `excludeUserFields` I see that they'll be excluded from `currentUser`, but if these fields on on the `User` model/table, how do you suggest protecting these from being leaked by a GraphQL query that requests then from `users` or `user`?

Yeah, as commented above I hadn't considered that as part of this implementation. On the one hand, as a developer it's 100% your responsibility to make sure your API is providing only the data you want to provide, but on the other, Redwood's default behavior is EXPOSE EVERYTHING when you generate your SDL. :grimacing: My first instinct is to update the generators to try and exclude those fields from the SDL generation. But I'm also curious to see what envelop can help us with!

This could be it's own standalone issue but: maybe we create some kind of `yarn rw audit` command that checks for things like exposed secure fields in your GraphQL, unsecure services, etc.

**Comment by cannikin** [comment]
> You may want to consider some additional claims here

The JWT isn't actually being used to verify anything, it's just (currently) the only way to transport information from the auth system back to the client, and is always added to GraphQL request headers (everything blows up otherwise). @peterp suggested we remove JWTs altogether for dbAuth as it's just an extra step we don't need. (That looked like it was going to touch a lot of things though so I'm waiting on his help before tackling that!)

You know that after that Netlify refresh token debacle there's no way on God's green earth I would ever add `exp` to anything. ;) That still doesn't work BTW, @dac09 even saw it happen on the Auth Playground with Netlify *after* the refresh token fix went live!

**Comment by cannikin** [comment]
## BEHOLD

Draft status rescinded—ready for review. Needs a little bit of Typescript help, but otherwise ready for review.

File: packages/api/src/functions/dbAuthErrors.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
+export const WrongVerbError = class extends Error {
```
Comment (dac09) [inline packages/api/src/functions/dbAuthErrors.js:1]: Rob, what do you think of directly exporting the class instead of assigning them to constants? e.g. -> 

```js
export class WrongVerbError extends Error {
  constructor(properVerb) {
    super(`Only accessible via ${properVerb}`)
    this.name = 'WrongVerbError'
  }
}
```

Comment (cannikin) [reply packages/api/src/functions/dbAuthErrors.js:1]: So many different ways to export things! Yeah that looks nicer.

**Comment by cannikin** [comment]
After the gracious help of @peterp tonight we've revamped the authentication mechanism between the web and api sides. Rather than creating a JWT we now just rely on the cookie itself to determine if you're logged in. There's a new decoder for dbAuth which decrypts the cookie and returns just the data inside of it (a simple object, `{id: 1234}`) and then it's up to you to look up and return your user in `getCurrentUser()` in `api/src/lib/auth.js` using that ID. This way you can scrub your own data, or join and include as much other data as you want, without relying on config options to get you close.

File: packages/api/src/auth/decoders/dbAuth.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+import { dbAuthSession } from '../../functions/dbAuth/shared'
+
+export const dbAuth = (
+  authHeaderValue: string,
+  req: { event: APIGatewayProxyEvent }
+) => {
+  const session = dbAuthSession(req.event)
+  const authHeaderUserId = authHeaderValue
+
+  if (session.id.toString() !== authHeaderUserId) {
```
Comment (peterp) [inline packages/api/src/auth/decoders/dbAuth.ts:12]: the `toString()` part is weird

Comment (peterp) [reply packages/api/src/auth/decoders/dbAuth.ts:12]: I assume TS, just a comment. Not an actionable thing.

Comment (cannikin) [reply packages/api/src/auth/decoders/dbAuth.ts:12]: SO that's because HTTP headers always come in as strings. But, the user's ID you get from the database could be a string (like a UUID) or could be an integer, depending on how it's stored. So this just makes sure it's always a string so that it can compare against he header.

Comment (cannikin) [reply packages/api/src/auth/decoders/dbAuth.ts:12]: And I don't know that this check is even required (since the session is encrypted, it's extremely unlikely that the wrong user ID is in there) but I figured since *something* is gonna go up in that auth header, we may as well double check that it's the correct value.

Comment (cannikin) [reply packages/api/src/auth/decoders/dbAuth.ts:12]: And that's just what I decided to put in that header, it could be anything, right? Right now it's the plain text of the user's ID, which some might consider a security risk (if it's an integer then you know you can start going through each of them in a row to try and hack something).

**Review by peterp — APPROVED** [review]
This looks good to me! <3

Mammoth PR!

**Comment by thedavidprice** [comment]
@peterp @dthyresson @cannikin 
_Note about possible future steps with Envelop via @redwoodjs/graphql-server package:_ 
There are a lot of new additions here to the /api package as well as direct import of 'aws-lambda'. So as is this won't be compatible with testing graphql-server (which might be the case for other Auth providers as well). Reminder that we'll need to evaluate refactor based on Envelop testing. And when we decouple the handler from aws-lambda, we'll need to resolve here as well.

Nothing to do at this time.

**Comment by thedavidprice** [comment]
🚀
